File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.cpp
#include <string>
#include <vector>

using namespace std;
int map[1001][1001];
int sum[1001][1001];
int solution(vector<vector<int>> board, vector<vector<int>> skill) {
    int answer = 0;
    for(int i = 0; i<board.size();i++)
        for(int j = 0; j<board[0].size(); j++)
            map[i][j] = board[i][j];
    
    for(auto sk : skill){
        int degree = (sk[0] == 1 ? -sk[5] : sk[5]);
        
        int r1 = sk[1], c1 = sk[2];
        int r2 = sk[3], c2 = sk[4];
        
        sum[r1][c1] += degree;
        sum[r1][c2 + 1] -= degree;
        sum[r2 + 1][c2 + 1] += degree;
        sum[r2 + 1][c1] -= degree;
        
    }
    for(int i = 0; i<board.size();i++)
        for(int j = 1; j<board[0].size(); j++)
            sum[i][j] += sum[i][j-1];
    
    for(int i = 1; i<board[0].size();i++)
        for(int j = 0; j<board.size(); j++)
            sum[i][j] += sum[i-1][j];
    

    
    for(int i = 0; i<board.size();i++)
        for(int j = 0; j<board[0].size(); j++)
            if(map[i][j] + sum[i][j]> 0)
                answer++;
            
    return answer;
}

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.java
class Solution {
    static int[][] mask;
    public int solution(int[][] board, int[][] skill) {
        int answer = 0;
        
        int n = board.length;
        int m = board[0].length;
        
        mask = new int[n + 1][m + 1];
        
        for(int[] s : skill){
            int x = (s[0] == 1 ? -s[5] : s[5]);
            
            mask[ s[1] ][ s[2] ] += x;
            mask[ s[1] ][ s[4] + 1 ] += -x;
            mask[ s[3] + 1 ][ s[2] ] += -x;
            mask[ s[3] + 1 ][ s[4] + 1] += x;
        }
        
        for(int i = 0; i < n; i++){ //행
            for(int j = 1; j < m; j++){
                mask[i][j] += mask[i][j - 1];
            }
        }
        
        for(int i = 1; i < n; i++){ //열
            for(int j = 0; j < m; j++){
                mask[i][j] += mask[i - 1][j];
            }
        }
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                board[i][j] += mask[i][j];
                if(board[i][j] > 0) answer++;
            }
        }
        
        return answer;
    }
}